README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# locateip

<!-- badges: start -->

[![R-CMD-check](https://github.com/clessn/locateip/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/clessn/locateip/actions/workflows/R-CMD-check.yaml)
[![Codecov test
coverage](https://codecov.io/gh/clessn/locateip/branch/main/graph/badge.svg)](https://app.codecov.io/gh/clessn/locateip?branch=main)
<!-- badges: end -->

The goal of locateip is to locate IP addresses using
[ip-api.com](https://ip-api.com/)

## Installation

For the most recent version, on CRAN:

``` r
install.packages("locateip")
```

You can install the development version of locateip from
[GitHub](https://github.com/):

``` r
# install.packages("devtools")
devtools::install_github("clessn/locateip")
```

## Example

``` r
library(locateip)

locate_ip("132.203.167.188", fields = c("status,message,country,city"))
#> # A tibble: 1 × 3
#>   status  country city  
#>   <chr>   <chr>   <chr> 
#> 1 success Canada  Québec
```

## Usage

[ip-api](https://ip-api.com/) is free for non-commercial use.

For API documentation and terms of service, see
[ip-api.com](https://ip-api.com/).


NEWS.md
# locateip (development version)

# locateip 0.1.3

* Added a `NEWS.md` file to track changes to the package.

* Added URL and bug report link to `DESCRIPTION`.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(locateip)

test_check("locateip")


tests/testthat/test-locate.R
test_that("tidy response works", {
  output <- tibble::tibble(
    status = "fail",
    message = "invalid query"
  )

  expect_equal(
    tidy_resp(response = "status,message\nfail,invalid query\n"),
    output
  )

  output <- tibble::tibble(
    status = "success",
    country = "Canada",
    city = "Fredericton",
    query = "0.0.0.0"
  )

  expect_equal(
    tidy_resp(response = "status,country,city,query\nsuccess,Canada,Fredericton,0.0.0.0\n"),
    output
  )
})

test_that("locate ip works", {
  output <- tibble::tibble(
    status = "success",
    country = "Canada",
    city = "Québec"
  )

  expect_equal(
    locate_ip("132.203.167.188", tidy = TRUE),
    output
  )

  output <- "status,country,city\nsuccess,Canada,Québec\n"

  expect_equal(
    locate_ip("132.203.167.188", tidy = FALSE),
    output
  )
})

test_that("create request works", {
  req <- create_req(ip = "132.203.167.188")
  expect_equal(
    req[["url"]],
    "http://ip-api.com/csv/132.203.167.188?fields=status%2Cmessage%2Ccountry%2Ccity"
  )

  req <- create_req(ip = NULL)
  expect_equal(req[["url"]],
               "http://ip-api.com/csv?fields=status%2Cmessage%2Ccountry%2Ccity")

  req <-
    create_req(ip = "132.203.167.188",
      fields = c(
        "status,message,country,countryCode,region,regionName,city,zip,lat,lon,timezone,isp,org,as,query"
      )
    )
  expect_equal(req[["url"]],
               "http://ip-api.com/csv/132.203.167.188?fields=status%2Cmessage%2Ccountry%2CcountryCode%2Cregion%2CregionName%2Ccity%2Czip%2Clat%2Clon%2Ctimezone%2Cisp%2Corg%2Cas%2Cquery")

  req <- create_req(ip = "132.203.167.188", lang = "en")
  expect_equal(req[["url"]],
               "http://ip-api.com/csv/132.203.167.188?fields=status%2Cmessage%2Ccountry%2Ccity&lang=en")

  req <- create_req(ip = "132.203.167.188", header = "true")
  expect_equal(req[["url"]],
               "http://ip-api.com/csv/132.203.167.188?fields=status%2Cmessage%2Ccountry%2Ccity&header=true")

  expect_error(get_location(ip = "132.203.167.188", lang = "uk"))

  # req <- create_req()
  # expect_equal(environment(req[["policies"]][["throttle_delay"]])[["rate"]], 0.75)
})


tests/testthat/test-validate.R
test_that("validate ipv6 works", {
  # Valid
  expect_equal(validate_ipv6("2a01:cb08:832a:500:d9e3:a1d0:3d20:5f7e"),
               TRUE)

  expect_equal(validate_ipv6("2001:0db8:0000:0000:0000:ff00:0042:8329"),
               TRUE)
  expect_equal(validate_ipv6("2001:db8::ff00:42:8329"),
               TRUE) # Abbreviation of previous

  # Not valid
  expect_equal(validate_ipv6("0"),
               FALSE)
  expect_equal(validate_ipv6(""),
               FALSE)
  expect_equal(validate_ipv6("NotAnIp"),
               FALSE)
})

test_that("validate ipv4 works", {
  # Valid
  expect_equal(validate_ipv4("132.203.167.188"),
               TRUE)
  expect_equal(validate_ipv4("0.0.0.0"),
               TRUE)
  expect_equal(validate_ipv4("127.255.255.255"),
               TRUE)
  expect_equal(validate_ipv4("128.0.0.0"),
               TRUE)
  expect_equal(validate_ipv4("191.255.255.255"),
               TRUE)
  expect_equal(validate_ipv4("192.0.0.0"),
               TRUE)
  expect_equal(validate_ipv4("223.255.255.255"),
               TRUE)

  # Not valid
  expect_equal(validate_ipv4("0.0.0"),
               FALSE)
  expect_equal(validate_ipv4("0"),
               FALSE)
  expect_equal(validate_ipv4(""),
               FALSE)
  expect_equal(validate_ipv4("NotAnIP"),
               FALSE)
})

test_that("validate ip works", {
  # Valid
  expect_equal(validate_ip("142.162.45.64"),
               TRUE)
  expect_equal(validate_ip("2a01:cb08:832a:500:d9e3:a1d0:3d20:5f7e"),
               TRUE)

  # Not valid
  expect_equal(validate_ip("0.0.0"),
               FALSE)
  expect_equal(validate_ip("0"),
               FALSE)
  expect_equal(validate_ip(""),
               FALSE)
  expect_equal(validate_ip("NotAnIP"),
               FALSE)
})


R/validate.R
#' Validate an IP address
#'
#' @description
#' `r lifecycle::badge("experimental")`
#'
#' Validate an IP adress as IPv4 or IPv6.
#'
#' @param ip String. IP address as IPv4 or IPv6.
#' @return Logical. TRUE if it's either IPv4 or IPv6, FALSE if not.
#' @noRd
validate_ip <- function(ip) {
  is_ipv4 <- validate_ipv4(ip)
  is_ipv6 <- validate_ipv6(ip)

  return(is_ipv4 | is_ipv6)
}

validate_ipv4 <- function(ip){
  valid <- grepl("^((25[0-5]|(2[0-4]|1\\d|[1-9]|)\\d)\\.?\\b){4}$", ip)
  return(valid)
}

validate_ipv6 <- function(ip){
  valid <- grepl("(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))", ip)
  return(valid)
}


R/locate_ip.R
#' Locate an IP adress
#'
#' `r lifecycle::badge('experimental')`
#'
#' For API documentation and terms of service, see [ip-api.com](https://ip-api.com/).
#'
#' @param ip A single IPv4/IPv6 address or a domain name. If you don't supply a query the current IP address will be used.
#' @param format Json, xml, csv, newline or php.
#' @return A response.
#' @export
#' @examples
#' resp <- locate_ip("142.162.45.64")
#'
#' resp |>
#'   httr2::resp_body_string()
locate_ip <- function(ip, format = "csv") {
  if (validate_ip(ip) == FALSE) {
    return(print("Pleade use a valid IP adress"))
  } else {
    resp <- httr2::request("http://ip-api.com") |>
      httr2::req_url_path_append(format) |>
      httr2::req_url_path_append(ip) |>
      httr2::req_user_agent("locateip (https://github.com/clessn; info@clessn.ca)") |>
      httr2::req_throttle(45 / 60) |>
      httr2::req_perform()

    return(resp)
  }
}

#' Validate an IP address
#'
#' @description
#' `r lifecycle::badge("experimental")`
#'
#' Internal function.
#'
#' Validate an IP adress as IPv4 or IPv6.
#'
#' @param ip String. IP address as IPv4 or IPv6.
#' @return Logical. TRUE if it's either IPv4 or IPv6, FALSE if not.
#' @describeIn validate Validate IPv4 or IPv6
validate_ip <- function(ip) {
  is_ipv4 <- validate_ipv4(ip)
  is_ipv6 <- validate_ipv6(ip)

  return(is_ipv4 | is_ipv6)
}

#' @describeIn validate Validate IPv4
validate_ipv4 <- function(ip){
  valid <- grepl("^((25[0-5]|(2[0-4]|1\\d|[1-9]|)\\d)\\.?\\b){4}$", ip)
  return(valid)
}

#' @describeIn validate Validate IPv6
validate_ipv6 <- function(ip){
  valid <- grepl("(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))", ip)
  return(valid)
}


R/locate.R
#' Locate an IP adress
#'
#' `r lifecycle::badge('experimental')`
#'
#' For API documentation and terms of service, see [ip-api.com](https://ip-api.com/).
#'
#' @param ip String. Either an IPv4 address, IPv6 address or a domain name. If NULL, will use the current IP address.
#' @param fields String. Response fields to pass on to the API.
#' @param lang String. Response language. An ISO 639 code supported by the API. Defaults to English.
#' @param ... Query parameters to pass on to the API.
#' @param tidy Logical. TRUE to return a tibble. FALSE to return a string.
#' @return A string or a tibble.
#' @export
#' @examples
#' locate_ip("132.203.167.188")
locate_ip <-
  function(ip = NULL,
           fields = c("status,message,country,city"),
           lang = "en",
           ...,
           tidy = TRUE) {
    resp <- get_location(ip, fields = fields, lang = lang, ..., format = "csv")

    string <- resp |>
      httr2::resp_body_string()

    if (tidy) {
      data <- tidy_resp(response = string)

      return(data)
    } else {
      return(string)
    }
  }

#' Create request for 'ip-api'
#'
#' `r lifecycle::badge('experimental')`
#'
#' For API documentation and terms of service, see [ip-api.com](https://ip-api.com/).
#'
#' @param ip String. Either an IPv4 address, IPv6 address or a domain name. If NULL, will use the current IP address.
#' @param fields String. Response fields to pass on to the API.
#' @param ... Query parameters to pass on to the API.
#' @param format String. Json, xml, csv, newline or php.
#' @return A response.
#'
#' @noRd
create_req <-
  function(ip = NULL,
           fields = c("status,message,country,city"),
           ...,
           format = "csv") {
    params <- list(fields = fields,
                   ...)

    req <- httr2::request("http://ip-api.com") |>
      httr2::req_url_path_append(format)

    if (is.null(ip) == FALSE) {
      req <- httr2::req_url_path_append(req, ip)
    }

    req <- req |>
      httr2::req_url_query(!!!params) |>
      httr2::req_user_agent("locateip (https://github.com/clessn; info@clessn.ca)") |>
      httr2::req_throttle(45 / 60)

    return(req)
  }

#' Get location
#'
#' `r lifecycle::badge('experimental')`
#'
#' For API documentation and terms of service, see [ip-api.com](https://ip-api.com/).
#'
#' @param ip String. Either an IPv4 address, IPv6 address or a domain name. If NULL, will use the current IP address.
#' @param fields String. Response fields to pass on to the API.
#' @param lang String. Response language. An ISO 639 code supported by the API. Defaults to English.
#' @param header Logical. Get field headers.
#' @param ... Query parameters to pass on to the API.
#' @param format String. Json, xml, csv, newline or php.
#' @return A response.
#'
#' @noRd
get_location <-
  function(ip = NULL,
           fields = c("status,message,country,city"),
           lang = "en",
           header = "true",
           ...,
           format = "csv") {
    lang <- match.arg(lang, c("en", "de", "es", "pt-BR", "fr", "ja", "zh-CN", "ru"))

    resp <- create_req(ip = ip, fields = fields, lang = lang, header = header, ..., format = format) |>
      httr2::req_perform()

    # If there's no remaining requests, pause execution
    if (as.numeric(httr2::resp_header(resp, "X-Rl")) == 0) {
      Sys.sleep(as.numeric(httr2::resp_header(resp, "X-Ttl")))
    }

    return(resp)
  }

#' Tidy response string into a tibble
#'
#' @description
#' `r lifecycle::badge('experimental')`
#'
#' Internal function.
#'
#' @param response Body string response with field headers
#' @noRd
#' @return Tibble.
tidy_resp <- function(response = NULL) {

  data <-readr::read_csv(response, show_col_types = FALSE)

  return(data)
}
